### PR TITLE
[API] Feature/job failure policy

### LINKED ISSUE #27997: Case report: Job somehow made 84k pods
Observed a case where a Job somehow created 84k pods. I think the user had set parallelism to 0 to pause it by the time I saw it. So I'm not sure how exactly to reproduce, but we should think about implementing some sanity limits. P1 since I was paged about this.

### LINKED ISSUE #30243: Maximum number of failures or failure backoff policy for Jobs
A max number of failures or failure backoff policy for Jobs would be useful.

Imagine you have an ETL job in production that's failing due to some pathological input.
By the time you spot it, it's already been rescheduled thousands of times. As an example, I had 20 broken jobs that kept getting rescheduled; killing them took forever -- and crashed the Kubernetes dashboard in the process.

Today, "restartPolicy" is not respected by Jobs because the goal is to achieve successful completion. (Strangely, "restartPolicy: Never" is still valid YAML.) This means failed jobs keep getting rescheduled. When you go to delete them, you have to delete all the pods they've been scheduled on. Deletes are rate limited, and in v1.3+, the verbose "you're being throttled" messages are hidden from you when you run the `kubectl` command to delete a job. So it just looks like it's taking forever! This is not a pleasant UX if you have a runaway job in prod or if you're testing out a new job and it takes minutes to clean up after a broken test.

What are your thoughts on specifying a maximum number of failures/retries or adding a failure backoff restart policy?

(cc @thockin , who [suggested I file this feature request](https://twitter.com/zazius/status/758816688171126784))

---

Ninja edit: [per this SO thread](http://stackoverflow.com/questions/35141607/is-there-a-max-retries-for-kubernetes-jobs), the throttling issue is avoidable by using the OnFailure restart policy to keep rescheduling to the same pods, rather than to new pods -- i.e. this prevents the explosion of the number of pods. And deadlines can help weed out failures after a certain amount of time.

However, suppose my ETL job takes an hour to run properly but may fail within seconds if the input data is bad. I'd rather specify a maximum number of retries than a high deadline.

### PR DESCRIPTION
**Pull request by clamoriniere1A**
**What this PR does / why we need it**: Implements the Backoff policy and failed pod limit defined in https://github.com/kubernetes/community/pull/583

**Which issue this PR fixes**: 
fixes #27997, fixes #30243

**Special notes for your reviewer**:
This is a WIP PR, I updated the api batchv1.JobSpec in order to prepare the backoff policy implementation in the JobController.

**Release note**:
```release-note
Add backoff policy and failed pod limit for a job
```

### CODE REVIEW

**Comment by kargakis** [comment]
@kubernetes/sig-apps-api-reviews

**Comment by clamoriniere1A** [comment]
CLA signed

**Comment by clamoriniere1A** [comment]
Hi @soltysh 

This PR is linked to the proposal “Backoff policy and failed pod limit”: https://github.com/kubernetes/community/pull/583

I started to update the jobcontroller.syncJob() logic, and I have 2 questions regarding the attended behaviours:
**1) Do we need to delete all the active running Pod in case the ```Backofflimit``` have been reached?** like it is done when we cross the “Job Deadine”.
(code link: https://github.com/clamoriniere1A/kubernetes/blob/feature/job_failure_policy/pkg/controller/job/jobcontroller.go#L521 )

**2) In order to handle properly the deletion of failed pods when the ```Job.Spec. FailedPodsLimit``` is reached.** 
I needed to add a new counter ```FailedAndDeleted``` in the ```Job.Status struct``` ( https://github.com/clamoriniere1A/kubernetes/blob/feature/job_failure_policy/pkg/apis/batch/types.go#L179 ). The new counter allows me to calculate the total number of failed Pod: “Failed and Active” + “Failed and deleted”.
**Do you think is a good approach?**

**Comment by soltysh** [comment]
@clamoriniere1A

> 1) Do we need to delete all the active running Pods in case the Backofflimit have been reached? like it is done when the Job cross the “Job Deadine”.

Yes. We need to stop whatever is already running. 

> 2) In order to handle properly the deletion of failed pods when the Job.Spec. FailedPodsLimit is reached.
> I needed to add a new counter FailedAndDeleted in the Job.Status struct ( https://github.com/clamoriniere1A/kubernetes/blob/feature/job_failure_policy/pkg/apis/batch/types.go#L179 ). The new counter allows me to calculate the total number of failed Pod: “Failed and Active” + “Failed and deleted”.
> Do you think it is a good approach?

I don't think it's needed. The `Job.Status.Failed` value should have all the pods that failed during the overall run. Not just currently available.

File: pkg/apis/batch/v1/defaults_test.go [hunk 1]
```diff
@@ -115,47 +123,95 @@ func TestSetDefaultJob(t *testing.T) {
...
+						ObjectMeta: metav1.ObjectMeta{Labels: defaultLabels},
+					},
+				},
+			},
+			expected: &batchv1.Job{
+				Spec: batchv1.JobSpec{
+					Completions:     newInt32(1),
+					Parallelism:     newInt32(1),
+					BackoffLimit:    newInt32(2),
+					FailedPodsLimit: newInt32(1),
```
Comment (soltysh) [inline pkg/apis/batch/v1/defaults_test.go]: That seems wrong, if you set it explicitly to 2, why you're expecting 1, here?

File: pkg/apis/batch/types.go [hunk 2]
```diff
@@ -160,6 +173,10 @@ type JobStatus struct {
 	// The number of pods which reached phase Failed.
 	// +optional
 	Failed int32
+
+	// The number of pods which reached phase Failed and deleted.
```
Comment (soltysh) [inline pkg/apis/batch/types.go]: Not needed, the above `Failed` field should hold information about all the failed pods. Currently it's the job controller that is broken, see https://github.com/kubernetes/kubernetes/issues/28486 for details.

**Review by soltysh — CHANGES_REQUESTED** [review]

File: pkg/apis/batch/types.go (hunk 2 above)
Comment (clamoriniere1A) [inline pkg/apis/batch/types.go]: Ok I will remove this field

File: pkg/apis/batch/v1/defaults_test.go (hunk 1 above)
Comment (clamoriniere1A) [inline pkg/apis/batch/v1/defaults_test.go]: Right, 
The fact is that I mess up some commits split. I did include in this commit the test logic update that is checking the value: BackoffLimit and FailedPodsLimit.

**Review by soltysh — APPROVED** [review]

**Comment by soltysh** [comment]
@erictune needs your sign-off for the API changes

**Comment by smarterclayton** [comment]
Based on reading of the proposal and conventions this is API approved.  The PR that enables failedpodslimit will still need API review, although I confirmed it matches the proposal.

/approve

**Comment by kargakis** [comment]
Is there a PR with the controller changes?

**Comment by cedriclam** [comment]
@kargakis, yes #51153 contains controller changes

**Comment by kargakis** [comment]
merci

**Pull request merged**